File: headers_list/src/main/java/CustomTypes/CustomDate.java
package CustomTypes;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.Objects;

public class CustomDate implements Comparable<CustomDate> {

    private final LocalDate date;
    private static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("dd/MM/yyyy");

    private CustomDate(LocalDate date) {
        this.date = date;
    }

    public static CustomDate of(int day, int month, int year) {
        return new CustomDate(LocalDate.of(year, month, day));
    }

    public static CustomDate parse(String dateString) {
        return new CustomDate(LocalDate.parse(dateString, FORMATTER));
    }

    public int getDay() {
        return date.getDayOfMonth();
    }

    public int getMonth() {
        return date.getMonthValue();
    }

    public int getYear() {
        return date.getYear();
    }

    public CustomDate addDays(int days) {
        return new CustomDate(this.date.plusDays(days));
    }

    public CustomDate subtractDays(int days) {
        return new CustomDate(this.date.minusDays(days));
    }

    @Override
    public int compareTo(CustomDate other) {
        return this.date.compareTo(other.date);
    }

    @Override
    public String toString() {
        return date.format(FORMATTER);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        CustomDate that = (CustomDate) o;
        return Objects.equals(date, that.date);
    }

    @Override
    public int hashCode() {
        return Objects.hash(date);
    }
}

File: headers_list/src/main/kotlin/CustomTypes/CustomDate.kt
package CustomTypes

import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.Objects

class CustomDate private constructor(private val date: LocalDate) : Comparable<CustomDate> {

    companion object {
        private val FORMATTER: DateTimeFormatter = DateTimeFormatter.ofPattern("dd/MM/yyyy")

        @JvmStatic
        fun of(day: Int, month: Int, year: Int): CustomDate {
            return CustomDate(LocalDate.of(year, month, day))
        }

        @JvmStatic
        fun parse(dateString: String): CustomDate {
            return CustomDate(LocalDate.parse(dateString, FORMATTER))
        }
    }

    fun getDay(): Int {
        return date.dayOfMonth
    }

    fun getMonth(): Int {
        return date.monthValue
    }

    fun getYear(): Int {
        return date.year
    }

    fun addDays(days: Int): CustomDate {
        return CustomDate(this.date.plusDays(days.toLong()))
    }

    fun subtractDays(days: Int): CustomDate {
        return CustomDate(this.date.minusDays(days.toLong()))
    }

    override fun compareTo(other: CustomDate): Int {
        return this.date.compareTo(other.date)
    }

    override fun toString(): String {
        return date.format(FORMATTER)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as CustomDate
        return date == that.date
    }

    override fun hashCode(): Int {
        return Objects.hash(date)
    }
}
